File: NeetCode/Binary_Search/Koko_Eating_Bananas/koko__eating__bananas.py
from typing import Optional, List
import math


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)

        while left < right:
            mid = (left + right) // 2
            total = 0
            for p in piles:
                total += (p + mid - 1) / mid
            if total <= h:
                right = mid
            else:
                left = mid + 1
        return left


if __name__ == "__main__":
    sol = Solution()

    piles = [25, 10, 23, 4]
    h = 4

    print(sol.minEatingSpeed(piles, h))
    print("Running Solution...")

File: NeetCode/Binary_Search/Koko_Eating_Bananas/Koko_Eating_Bananas.cpp
// Koko_Eating_Bananas.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int minEatingSpeed(vector<int> &piles, int h)
    {
        // find the max value
        auto rIdx = max_element(piles.begin(), piles.end());
        int right = rIdx - piles.begin();
        right = piles[right];

        int left{1}, mid{-1}, total{-1};
        // use binary search to find the min
        while (left < right)
        {
            mid = (left + right) / 2;
            total = 0;
            // check to how many are consumed with that banana /hr
            for (int p : piles)
            {
                total += (p + mid - 1) / mid;
            }

            if (total <= h)
            {
                right = mid;
            }
            else
            {
                left = mid + 1;
            }
        }
        return left;
    }
};
